Feat/Currency Conversion

mickey4653 opened the pull request:
# Bitnob API Integration

## Overview
This PR implements a Node.js API service that integrates with the Bitnob API to provide currency conversion and exchange rate functionality. The implementation follows a clean architecture pattern with proper separation of concerns and robust error handling.

## Features Implemented

### 1. Currency Conversion
- **Endpoint:** `POST /api/currency/convert`
- Converts between different currencies using Bitnob's exchange rates
- Input validation using Zod schema
- Supports conversion between any supported currency pairs

### 2. Exchange Rate Information
- **Endpoint:** `GET /api/currency/rates`
- Retrieves all available exchange rates from Bitnob
- Returns comprehensive rate information for all supported currencies

### 3. Specific Currency Rate
- **Endpoint:** `GET /api/currency/rate/:currency`
- Retrieves exchange rate information for a specific currency
- Supports any currency code supported by Bitnob

## Technical Implementation Details

### Architecture
- Follows MVC pattern with clear separation of concerns:
  - **Controllers:** Handle request/response logic
  - **Services:** Implement business logic and API integration
  - **Routes:** Define API endpoints and routing

### Key Components

#### Currency Service
- Implements Bitnob API integration
- Handles all currency-related operations
- Includes comprehensive error handling and logging
- Uses Axios for HTTP requests

#### Input Validation
- Implements Zod schema validation
- Validates currency codes and amounts
- Provides detailed error messages

#### Error Handling
- Global error handling middleware
- Detailed error responses
- Proper HTTP status codes
- Comprehensive error logging

## Dependencies
- Express.js for API framework
- Axios for HTTP requests
- Zod for input validation
- Dotenv for environment configuration
- Jest for testing
- ESLint for code quality

## Testing
- Jest configuration set up
- Test directory structure in place
- Ready for implementing unit and integration tests

## Security
- API key management through environment variables
- Input validation and sanitization
- Proper error handling to prevent information leakage

## Documentation
- Clear code comments
- `README.md` with setup instructions
- API endpoint documentation

mickey4653 committed:
Feat:
	- added currency conversion routes, controller, services
	- work in progress

mickey4653 committed:
Feat:
	- added currency conversion correct endpoints
	- all endpoints are working now

mickey4653 committed:
Test:
	- still working on writing tests

mickey4653 committed:
Fix/Test:
	- all tests are passing successfully

ngodi commented on src/services/currencyService.js:
```diff
@@ -0,0 +1,172 @@
+import axios from 'axios';
+import dotenv from 'dotenv';
+
+dotenv.config();
+
+const BITNOB_API_URL = 'https://sandboxapi.bitnob.co/api/v1';
+const API_KEY = process.env.BITNOB_API_KEY;
```
better in the .env and handled in the config class already

ngodi commented on src/services/currencyService.js:
```diff
@@ -0,0 +1,172 @@
+        }
+      });
+
+      const response = await this.client.get(`/wallets/payout/rate/${currency}`);
```
could you please extend from the base bitnob class, no need to initialize the api-keys again

ngodi reviewed (changes requested):
Please review suggestions

ngodi committed:
Merge branch 'dev' into feat/currency-conversion

mickey4653 committed:
Refactor:✅
	- resolved all lint errors
	- created BaseBitnobService class so it can reused in currencyService
	- removed duplicated code

mickey4653 committed:
Resolved Conflicts
Merge branch 'feat/currency-conversion' of https://github.com/internPulse-cohort8/bitnob-api-nodejs into feat/currency-conversion

mickey4653 committed:
Fix:
	- Duplicate imports and app initialization
	- Inconsistent port configuration
	- Missing error handling middleware
	- Missing currency routes

mickey4653 committed:
Fix:
	- removed lint errors
